<a href="https://colab.research.google.com/github/rqchao/INFO159/blob/main/HW3/HW_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Homework 3: Language Models, Contextual Embedding and BERT

In this homework, we will explore implementations of various language models we saw in lecture. We will explore BERT and measure perplexity. 

##Set Up

If you're opening this Notebook on colab, you will probably need to install Transformers. Make sure your version of Transformers is at least 4.11.0

In [1]:
! pip install transformers

In [2]:
import transformers
print(transformers.__version__)

c:\Users\chaoq\anaconda3\envs\nlp\lib\site-packages\tqdm\auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


4.24.0


IMPORTANT: For this assignment, GPU is not necessary. The following code block should show "Running on cpu". 
Go to Runtime > Change runtime type > Hardware accelerator > None if otherwise.

In [3]:
import torch
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Running on {}".format(device))

Running on cpu


# Masking

One of the core ideas to wrap your head around with transformer-based language models (and PyTorch) is the concept of *masking*---preventing a model from seeing specific tokens in the input during training.

* BERT training relies on the concept of *masked language modeling*: masking a random set of input tokens in a sequence and attempting to predict them.  Remember that BERT is *bidirectional*, so that it can use all of the other non-masked tokens in a sentence to make that prediction.

* The GPT class of models acts as a traditional left-to-right language model (sometimes called a "causal" LM) .  This family also uses self-attention based transformers---but, when making a prediction for the word $w_i$ at position $i$, it can only use information about words $w_1, \ldots, w_{i-1}$ to do so.  All of the other tokens following position $i-1$ must be *masked* (hidden from view).


Think about a mask as a matrix that's applied to every input $w$ when generating an output $o$ that determines whether an given $o_i$ is allowed to access each token in $w$.  For example, when passing a three-word input sequence through a transformer (to yield a three-word output sequence), a mask is a $3 \times 3$ matrix where the cells are essentially  answering the following questions:

\begin{bmatrix}
o_1 \; \textrm{hide} \; w_1\textrm{?} & o_1 \; \textrm{hide} \; w_2\textrm{?} & o_1 \; \textrm{hide} \; w_3\textrm{?} \\
o_2 \; \textrm{hide} \; w_1\textrm{?} & o_2 \; \textrm{hide} \; w_2\textrm{?} & o_2 \; \textrm{hide} \; w_3\textrm{?} \\
o_3 \; \textrm{hide} \; w_1\textrm{?} & o_3 \; \textrm{hide} \; w_2\textrm{?} & o_3 \; \textrm{hide} \; w_3\textrm{?} \\
\end{bmatrix}

In the masks we will consider below, 1 denotes that a position should be hidden; 0 denotes that it should be visible. Consider this mask:

\begin{bmatrix}
0 & 1 & 1 \\
1 & 0 & 1 \\
1 & 1 & 0
\end{bmatrix}

And consider this sequence:

\begin{bmatrix}
\textrm{John} & \textrm{likes}  & \textrm{dogs}  \\
\end{bmatrix}

When applying this mask to that sequence, we're saying that when we're generating the output for $o_1$ (*John*), we can only consider $w_1$ as an input (*John*).  Likewise, when we generate the output for $o_2$ (*likes*), we can only consider $w_2$ as an input (*likes*), and so on.  (This is a terrible mask!  But illustrates what function a mask performs.)

The following code illustrates how this works for that particular mask.


In [198]:
import numpy as np

def visualize_masking(sequences, mask):
  
  print(mask)
  for sequence in sequences:
    for i in range(len(sequence)):
      visible=[]
      for j in range(len(sequence)):
        if mask[i][j]==0:
          visible.append(sequence[j])
      print("for word %s, the following tokens are visible: %s" % (sequence[i], visible))
    print()

In [199]:
sequences=[["This", "is", "a", "sentence", "that", "has", "exactly", "ten", "tokens", "."], ["Here's", "another", "sequence", "with", "10", "words", "like", "the", "last", "."]]	

seq_length=len(sequences[0])

test_mask=np.ones((seq_length,seq_length))
for i in range(seq_length):
  test_mask[i,i]=0

visualize_masking(sequences, test_mask)



[[0. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 0. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 0. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 0. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 0. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 0. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 0. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 0.]]
for word This, the following tokens are visible: ['This']
for word is, the following tokens are visible: ['is']
for word a, the following tokens are visible: ['a']
for word sentence, the following tokens are visible: ['sentence']
for word that, the following tokens are visible: ['that']
for word has, the following tokens are visible: ['has']
for word exactly, the following tokens are visible: ['exactly']
for word ten, the following tokens are visible: ['ten']
for word tokens, the following tokens are visible: ['tokens']
for word ., the following tokens are visible: ['.']

for word Here's, the following tokens are visible: ["Here's"]
for word another, the follow

##Q1.  
As we discussed in class, BERT masks a random set of words in the input and attempts to reconstruct those words as output.  Create a mask that randomly masks token positions 2 and 7 (for an input sequence length of 10 tokens, with 0 being the position of the first token).  For an input sequence of 10 tokens, you should generate output representations for all 10 tokens (i.e., $[o_1, \ldots, o_{10}]$ in the notation above, but each representation must ignore the same 2 input tokens.

In [200]:
def create_bert_mask(seq_length):
  mask=np.zeros((seq_length,seq_length))
  # implement BERT mask here
  
  # BEGIN SOLUTION
  for i in range(seq_length):
    mask[i,1]=1
    mask[i,6]=1
  # END SOLUTION

  return mask


##Q2
A left-to-right language model (such as GPT) can only use information from input words $[w_1, \ldots, w_{i}]$ when generating the representation for output $o_i$.  Encode this as a mask as well.

In [201]:
def create_causal_mask(seq_length):
  mask=np.zeros((seq_length,seq_length))
  # implement causal mask here

  # BEGIN SOLUTION
  for i in range(seq_length):
    for j in range(i+1,seq_length):
      mask[i,j]=1
  # END SOLUTION
  sequences=[["This", "is", "a", "sentence", "that", "has", "exactly", "ten", "tokens", "."], ["Here's", "another", "sequence", "with", "10", "words", "like", "the", "last", "."]]	
  visualize_masking(sequences, mask)
  
  return mask


Now let's go ahead and embed these masks within a model.  First, we'll load some textual data (from Austen's *Pride and Prejudice*).

In [202]:
!wget https://www.gutenberg.org/files/1342/1342-0.txt

'wget' is not recognized as an internal or external command,
operable program or batch file.


In [203]:
import nltk
from nltk import word_tokenize
from collections import Counter

In [204]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\chaoq\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

Let's read in the data and tokenize it; for this homework, we'll only work with the first 10,000 tokens of that book; we'll keep only the most frequent 1,000 word types (all other tokens will be mapped to an [UNK] token).

In [205]:
def read_data(filename):
  with open(filename) as file:
    data=file.read().lower()
    first10K=' '.join(data.split(" ")[:10000])
    toks=nltk.word_tokenize(first10K)[:10000]
    vocab={"[PAD]":0, "[UNK]":1}
    counts=Counter()
    for tok in toks:
      counts[tok]+=1
    for v, _ in counts.most_common(1000):
      vocab[v]=len(vocab)
    tokids=[]
    for tok in toks:
      tokid=1
      if tok in vocab:
        tokid=vocab[tok]
      
      tokids.append(tokid)

    return tokids, vocab  

Now let's specify our model in PyTorch.

In [206]:
from torch import nn
import torch

class MaskedLM(nn.Module):
    def __init__(self, vocab, mask, d_model=512):       
        super().__init__()
        self.vocab=vocab
        self.mask=mask
        vocab_size=len(vocab)
        self.embeddings=nn.Embedding(1002,512)
        encoder_layer = nn.TransformerEncoderLayer(d_model, nhead=8, batch_first=True)
        self.transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=6)
        self.linear=torch.nn.Linear(d_model, vocab_size)
        self.rev_vocab={vocab[k]:k for k in vocab}

    def forward(self, input): 
        # first we pass the input word IDS through an embedding layer to get embeddings for them
        input=self.embeddings(input)
        # then we pass those embeddings through a transformer to get contextual representations, masking the input where appropriate
        out = self.transformer_encoder.forward(input, mask=self.mask)        
        # finally we pass those embeddings through a linear layer to transform it into the output space (the size of our vocabulary)
        h=self.linear(out)
        return h

In [207]:
def get_batches(xs, ys, batch_size=32):
    batch_x=[]
    batch_y=[]
    for i in range(0, len(xs), batch_size):
        batch_x.append(torch.LongTensor(xs[i:i+batch_size]).to(device))
        batch_y.append(torch.LongTensor(ys[i:i+batch_size]).to(device))
    return batch_x, batch_y

In [208]:
tokids, vocab=read_data("1342-0.txt")  

In [209]:
def train(mask, data_function, tokids, vocab):

    mask=torch.BoolTensor(mask).to(device)

    num_labels=len(vocab)
    model=MaskedLM(vocab, mask).to(device)
    optimizer=torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=1e-5)
    cross_entropy=nn.CrossEntropyLoss()
    losses=[]

    xs, ys=data_function(tokids)

    batch_x, batch_y=get_batches(xs, ys)

    for epoch in range(1):
        model.train()
        
        for x, y in list(zip(batch_x, batch_y)):
            x, y = x.to(device), y.to(device)
            y_pred=model.forward(x)
            loss=cross_entropy(y_pred.view(-1, num_labels), y.view(-1))
            losses.append(loss.item())
            print(loss)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

Our model and training process are now all defined; all that remains is to pass our inputs and outputs through it to train.  Your job here is to create the correct inputs (x) and outputs (y) to train a left-to-right (causal) language model.

##Q3
Write a function that takes in a sequence of token ids $[w_1, \ldots, w_n]$ and segments it into 8-token chunks -- e.g., $x_1=[w_1, \ldots, w_8]$, $x_2=[w_9, \ldots, w_{16}]$, etc.  For each $x_i$, also create its corresponding $y_i$.  Given this language modeling specification, each $y_i$ should also contain 8 values (for each token in $x_i$).  Keep in mind this is a left-to-right causal language model; your job is to figure out the values of y that respects this design.  At token position $i$, when a model has access to $[w_1, \ldots, w_i]$, which is the true $y_i$ for that position? Each element in $y$ should be a word ID (i.e., an integer).

In [210]:
def get_causal_xy(data, max_len=8):
    xs=[]
    ys=[]
    # BEGIN SOLUTION
    xs = [data[i:i+max_len] for i in range(0, len(data), max_len)]
    ys = [data[i+1:i+max_len+1] for i in range(0, len(data), max_len)]
    
    if len(xs[-1]) < max_len:
      xs = xs[:-1]
      ys = ys[:-1]
    return xs, ys

In [211]:
seq_length=8

train(create_causal_mask(seq_length=seq_length), get_causal_xy, tokids, vocab)

tensor(7.0310, grad_fn=<NllLossBackward0>)
tensor(6.4619, grad_fn=<NllLossBackward0>)
tensor(6.3741, grad_fn=<NllLossBackward0>)
tensor(6.1574, grad_fn=<NllLossBackward0>)
tensor(6.2277, grad_fn=<NllLossBackward0>)
tensor(6.0824, grad_fn=<NllLossBackward0>)
tensor(5.9969, grad_fn=<NllLossBackward0>)
tensor(6.0274, grad_fn=<NllLossBackward0>)
tensor(6.1891, grad_fn=<NllLossBackward0>)
tensor(6.5385, grad_fn=<NllLossBackward0>)
tensor(6.2831, grad_fn=<NllLossBackward0>)
tensor(5.6186, grad_fn=<NllLossBackward0>)
tensor(5.4228, grad_fn=<NllLossBackward0>)
tensor(5.2465, grad_fn=<NllLossBackward0>)
tensor(5.0395, grad_fn=<NllLossBackward0>)
tensor(5.1205, grad_fn=<NllLossBackward0>)
tensor(5.0118, grad_fn=<NllLossBackward0>)
tensor(4.8325, grad_fn=<NllLossBackward0>)
tensor(4.7379, grad_fn=<NllLossBackward0>)
tensor(5.2983, grad_fn=<NllLossBackward0>)
tensor(5.3339, grad_fn=<NllLossBackward0>)
tensor(5.2319, grad_fn=<NllLossBackward0>)
tensor(4.5574, grad_fn=<NllLossBackward0>)
tensor(4.92

##Q4 (Write-up)  
In this model, as implemented, does the following equivalence hold?

$$
P(y_4 \mid w_1 = \textrm{go}, w_2=\textrm{ahead}, w_3=\textrm{make}, w_4=\textrm{my})= P(y_4 \mid w_1 = \textrm{ahead}, w_2=\textrm{my}, w_3=\textrm{make}, w_4=\textrm{go})
$$

Why or why not?

# Perplexity
To evaluate how good our language model is, we use a metric called perplexity. The perplexity of a language model (PP) on a test set is the inverse probability of the test set, normalized by the number of words. Let $W = w_{1}w_{2}\dots w_{N}$. Then,

$$PP(W) = \sqrt[N]{\prod_{i = 1}^{N}\frac{1}{P(w_{i}|w_{1}\dots w_{i - 1})}}$$

However, since these probabilities are often small, taking the inverse and multiplying can be numerically unstable, so we often first compute these values in the log domain and then convert back. So this equation looks like:

$$\ln PP(W) = \frac{1}{N} \sum_{i = 1}^{N} -\ln P(w_{i}|w_{1}\dots w_{i - 1})$$

$$\implies PP(W) = e^{\frac{1}{N} \sum_{i = 1}^{N} -\ln P(w_{i}|w_{1}\dots w_{i - 1})}$$

Here we want to calculate the perplexity of [pretrained BERT model](https://huggingface.co/bert-base-uncased) on text from different sources. When calculating perplexity with BERT, we'll use a related measure of pseudo-perplexity, which allow us to condition on the bidirectional context (and not just the left context, as in standard perplexity):

$$PP(W) = e^{\frac{1}{N} \sum_{i = 1}^{N} -\ln P(w_{i} \mid w_{1}\dots w_{i - 1}, w_{i+1}, \ldots, w_n)}$$


First, let's instantiate a BERT model, along with its WordPiece tokenizer.

In [212]:
from transformers import AutoModelForMaskedLM, AutoTokenizer
import torch
import numpy as np

model_name = 'bert-base-uncased'
model = AutoModelForMaskedLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)
model=model.to(device)

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


Let's see how the BERT tokenizer tokenizes a sentence into a sequence of WordPiece ids.  Note how BERT tokenization automatically wraps an input sentences with [CLS] and [SEP] tags.

In [213]:
sentence = "A dog landed on Mars"
tensor_input = tokenizer(sentence, return_tensors="pt")
print(tensor_input)
tensor_input_ids = tensor_input["input_ids"]
print(tensor_input_ids)
print(tokenizer.convert_ids_to_tokens(tensor_input_ids[0]))

{'input_ids': tensor([[ 101, 1037, 3899, 5565, 2006, 7733,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1]])}
tensor([[ 101, 1037, 3899, 5565, 2006, 7733,  102]])
['[CLS]', 'a', 'dog', 'landed', 'on', 'mars', '[SEP]']


Now let's see how we can calculate output probabilities using this model.  The output of each token position $i$ gives us $P(w_i \mid w_1, \ldots, w_n)$---the probability of the word at that position over our vocabulary, given *all* of the words in the sentence.

In [214]:
with torch.no_grad():
  output = model(tensor_input_ids)
  logits = output.logits
  # logits here are the unnormalized scores, so let's pass them through the softmax 
  # to get a probability distribution
  softmax = torch.nn.functional.softmax(logits, dim = -1)
  # for one input sequence, the shape of the resulting distribution is: 
  # 1 x [length of input, in WordPiece tokens] x (the size of the BERT vocabulary)
  print(softmax.shape) # [1, 7, 30522]
  input_ints=tensor_input_ids.numpy()[0]
  # Let's print the probability of the true inputs
  wp_tokens=tokenizer.convert_ids_to_tokens(input_ints)
  for i in range(len(input_ints)):
    prob=softmax[0][i][input_ints[i]].numpy()
    print("%s\t%s\t%.5f" % (wp_tokens[i], input_ints[i], prob))

torch.Size([1, 7, 30522])
[CLS]	101	0.00000
a	1037	0.99281
dog	3899	0.99052
landed	5565	0.99809
on	2006	0.99874
mars	7733	0.00133
[SEP]	102	0.00000


Note that $w_i$ is in the range $[w_1, \ldots, w_n]$ -- clearly the probability of a word is going to be high when we can observe it in the input! Let's do some masking to calculate $P(w_i \mid w_1, \ldots w_{i-1}, w_{i+1}, w_n)$.  Now annoyingly, BERT's `attention_mask` function only works for padding tokens; to mask input tokens, we need to intervene in the input and replace a WordPiece token that we're predicting with a special [MASK] token (BERT tokenizer word id `103`).

In [215]:
import copy

with torch.no_grad():
  # let's make a copy of the original word ids so we can mask one of the tokens
  masked_input_ids=copy.deepcopy(tensor_input_ids)
  # we'll mask the second word
  masked_input_ids[0][1]=tokenizer.convert_tokens_to_ids("[MASK]")

  print("The second word here now is [MASK] token ID '103': ", masked_input_ids)

  # now let's run that through BERT in the same way we did before
  output = model(masked_input_ids)
  logits = output.logits

  softmax = torch.nn.functional.softmax(logits, dim = -1)
  input_ints=tensor_input_ids.numpy()[0]

  wp_tokens=tokenizer.convert_ids_to_tokens(input_ints)
  i=1
  prob=softmax[0][i][input_ints[i]].numpy()
  print("%s\t%s\t%.5f" % (wp_tokens[i], input_ints[i], prob))

The second word here now is [MASK] token ID '103':  tensor([[ 101,  103, 3899, 5565, 2006, 7733,  102]])
a	1037	0.13965


You can see the probability of "a" as the second token has gone down to 0.13965 when we mask it.  This is the $P(w_1 =\textrm{a} \mid w_0, w_2, \ldots, w_n)$.  At this point you should have everything you need to calculate the BERT pseudo-perplexity of an input sentence.

##Q5
Implement the pseudo-perplexity measure described above, calculating the perplexity for a given model, tokenizer, and sentence. 

The function calculates the average probability of each token in the sentence given all the other tokens. We need to predict the probability of each word in a sentence by masking the one word to predict. Note that you should not include the probabilities of the [CLS] and [SEP] tokens in your perplexity equation -- those tokens are not part of the original test sentence.



In [216]:
# This function calculates the perplexity of a language model, given a sentence and its corresponding tokenizer

# Inputs:
# model: language model being used to calculate the perplexity
# tokenizer: tokenizer that is used to preprocess the input sentence
# sentence: input sentence string for which perplexity is to be calculated

# Outputs:
# returns perplexity of the input sentence

def perplexity(model, tokenizer, sentence):

  # hints: you'll need to:
  # encode the input sentence using the tokenizer
  # for each WordPiece token in the sentence (except [CLS] and [SEP]), mask that single token and 
  # calculate the probability of that true word at the masked position
  # don't calculate perplexity for the [CLS] and [SEP] tokens (which are not part of the original test sentence).
  perplexity=None
  # BEGIN SOLUTION
  perplexity = 0
  tensor_input = tokenizer(sentence, return_tensors="pt")
  tensor_input_ids = tensor_input["input_ids"]
  input_len=len(tensor_input_ids.numpy()[0])
  # Loop over each token in the sentence
  for i in range(1, input_len - 1):
    # Mask the current token
    masked_input_ids=copy.deepcopy(tensor_input_ids)
    masked_input_ids[0][i]=tokenizer.convert_tokens_to_ids("[MASK]")
    output = model(masked_input_ids)
    logits = output.logits

    softmax = torch.nn.functional.softmax(logits, dim = -1)
    input_ints=tensor_input_ids.numpy()[0]
    prob=softmax[0][i][input_ints[i]].detach().numpy()
    log_prob = np.log(prob)
    perplexity -= log_prob
  # Calculate the average perplexity
  perplexity /= (input_len - 2)
  # Convert the average perplexity from log space to perplexity space
  perplexity = np.exp(perplexity)
  # END SOLUTION

  return perplexity

In [217]:
print(perplexity(sentence='London is the capital of the United Kingdom.', model=model, tokenizer=tokenizer))

1.0596667966797595


# No credit.  

We provide [texts](https://people.ischool.berkeley.edu/~dbamman/text_from_different_sources.txt) from 4 different sources ([Wikipedia](https://www.kaggle.com/datasets/jrobischon/wikipedia-movie-plots), [Yelp](https://www.kaggle.com/datasets/omkarsabnis/yelp-reviews-dataset), [Fiction](https://github.com/dbamman/litbank), [Twitter](https://github.com/dbamman/anlp21/blob/main/data/potus_tweets.json)) collected from open-source datasets. Each category has 125 entries.

In [ ]:
!wget https://people.ischool.berkeley.edu/~dbamman/text_from_different_sources.txt

--2023-02-11 04:41:01--  https://people.ischool.berkeley.edu/~dbamman/text_from_different_sources.txt
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 61117 (60K) [text/plain]
Saving to: ‘text_from_different_sources.txt’

text_from_different 100%[===================>]  59.68K  --.-KB/s    in 0.03s   

2023-02-11 04:41:02 (2.02 MB/s) - ‘text_from_different_sources.txt’ saved [61117/61117]



In [ ]:
text_by_genre={}
with open('text_from_different_sources.txt') as file:
  file.readline()
  for line in file:
    cols=line.rstrip().split("\t")
    genre=cols[0]
    text=cols[1]

    if genre not in text_by_genre:
      text_by_genre[genre]=[]
    text_by_genre[genre].append(text)

for genre in text_by_genre:
  print(genre, len(text_by_genre[genre]))

Wikipedia 125
Yelp 125
Fiction 125
Twitter 125


Calculate perplexity on each genre over all of the words present within it; each line contains exactly one sentence for each genre.

The output perplexity_by_genre = {} is a dictionary mapping genre to a list of perplexities for each sentence in that genre. For computational purpose, we only take the first 25 sentences as an example (still this can take up to 10 minutes to run), feel free to change 25 to smaller numbers.

e.g. perplexity_by_genre['Wikipedia'] should be a list of 25 perplexities (one for each Wikipedia row in the input file).

In [ ]:
import numpy as np
def calculate_perplexity_by_genre(text_by_genre):
  perplexity_by_genre = {}
  for genre in text_by_genre:
    perplexity_by_genre[genre] = []
    for text in text_by_genre[genre][:25]: # change 25 to smaller numbers if necessary
      p = perplexity(sentence=text, model=model, tokenizer=tokenizer)
      perplexity_by_genre[genre].append(p)
  return perplexity_by_genre

In [ ]:
# running this might take up to 10 minutes
perplexity_by_genre = calculate_perplexity_by_genre(text_by_genre)
for genre in perplexity_by_genre:
  print("Genre:",genre,", mean perplexity:",np.mean(perplexity_by_genre[genre]))

##Question: 
What do you think are the reasons for the wide variation in perplexity of different categories of corpus? (hint: think about the training data of the pre-trained BERT model)


Which of these is a true language model, and why?

In [1]:
"""
objective: Determine if a list of 5-element tuples represents a sentence with 'projective' dependency tree

params: tokens is a list of (idd, tok, pos, head, lab) for a sentence
return True if and only if the sentence has a projective dependency tree
"""
# instructions
# there is more than one way to carry this out correctly. 
# We recommend the following procedure:
# Part 1
# Iterate over the tuples and store their relevant components a dictionary
# with the tuples' id ("idd") as the key and its head ("head") as the value
# this dictionary should thus contain dependency relations in the form {idd: head}
# Part 2
# For every dependency relation in that dictionary, extract the head
# establish the left and right bounds for your dependency-path search, with
# the left bound being: 1 + the smaller of (head, idd) and 
# the right bound, conversely, being the bigger of (head, idd)
# Part 3
# Check if every word in the tree is reachable by following the path of dependencies
# complete the reachable method

# hints
# -the left bound gets a 1 added to it because you don't need to check for crossing edges 
#   when you have a token that's directly besides its head
# - a token with '0' value for its head denotes that it is the root of the sentence
# -if any word in the tree is not reachable, this means there is non-projectivity
#   which is when your method needs to return False.
# -otherwise, return True!
# -you are welcome to remove any and all of our starter code and create your own logic for this
# if you do so, please *do not* remove the BEGIN / END Solution flags!
def is_projective(tokens):

    # BEGIN SOLUTION
    def reachable(i, head, heads):
        if i == head:
            return True
        elif i not in heads:
            return False
        else:
            return reachable(heads[i], head, heads)
    
    heads = {} # populate this (Part 1)
    for idd, tok, pos, head, lab in tokens:
        heads[idd] = head

    for dep_to_check in heads: # populate this (Part 2)
        head = heads[dep_to_check]
        left = min(head, dep_to_check)
        right = max(head, dep_to_check)
        for i in range(left + 1, right):
          if not reachable(i, head, heads):
            return False
            
    return True
    # END SOLUTION

In [2]:
"""
objective: perform the SHIFT operation

params:
    - wbuffer: input buffer
    - stack: what we're buliding our parse on
    - arcs: our dependency parse for a single token (containing the dependency relation, head-token ID, its own token ID)
    - configurations: state of the parse at a given decision point
    - gold_transitions: the operations we're applying at each decision point
"""

# instructions
# 1. update configurations
# 2. update gold_transitions
# 3. remove word from front of buffer and push it onto stack
# hints
# -this can be completed in just a few lines.
# -we have provided the code for updating configurations, so you just need to
# update gold_transitions, then the stack and wbuffer accordingly
# -since this is a SHIFT operation, arcs does not come into play because no head-dependent relationships are being asserted
# -the .pop() operation removes an item at the given index from a list
# defaulting to the last element in the list when no index is provided

def perform_shift(wbuffer, stack, arcs, configurations, gold_transitions):    
    
    # BEGIN SOLUTION
    # update configurations (Part 1)
    configurations.append((list(wbuffer), list(stack), list(arcs)))

    # update gold_transitions (Part 2)
    gold_transitions.append('SHIFT')

    # remove word from front of buffer and push it onto stack (Part 3)
    stack.append(wbuffer.pop())
    
    # END SOLUTION

In [3]:
"""
objective: perform LEFTARC_ and RIGHTARC_ operations

params:
    - direction: {"LEFT", "RIGHT"}
    - dep_label: label for the dependency relations
    - wbuffer: input buffer
    - stack: what we're buliding our parse on
    - arcs: our dependency parse for a single token (containing the dependency relation, head-token ID, its own token ID)
    - configurations: state of the parse at a given decision point
    - gold_transitions: the operations we're applying at each decision point
"""

# instructions
# 1. update configurations
# 2. update transitions
# 3. use the first 2 elements on the stack to create a (head, child) pairing, based on the type of shift we're doing
# 4. update arcs
# 5. update the stack
# hints
# -updating configurations is identical to the perform_shift function
# -be sure the information you're inserting into gold_transitions is formatted properly (see above)
# -be sure that the orientation of the (head,child) pairing is correct

def perform_arc(direction, dep_label, wbuffer, stack, arcs, configurations, gold_transitions):

    # update configurations (Part 1)
    configurations.append((list(wbuffer), list(stack), list(arcs)))

    # update transitions (Part 2)
    gold_transitions.append(f"{direction}ARC_{dep_label}")

    # setup head and child items (Part 3)
    if direction == "LEFT":
        head = stack[-1]
        child = stack[-2]
    elif direction == "RIGHT":
        head = stack[-2]
        child = stack[-1]

    # update arcs (Part 4)
    arcs.append((dep_label, head, child))

    # update stack (Part 5)
    if direction == "LEFT":
        stack.pop(-2)
    elif direction == "RIGHT":
        stack.pop()

    return wbuffer, stack, arcs, configurations, gold_transitions


In [34]:
"""
objective: based on inputs, return the correct configurations and actions for the parser.

params:
wbuffer: a list of word indices; the top of buffer is at the end of the list
stack: a list of word indices; the top of buffer is at the end of the list
arcs: a list of (label, head, dependent) tuples
deps: the existing dependency tree

Return configurations and gold_transitions
"""

# instructions
# Initialize return lists
# Check for simple cases (that we've translated the tree entirely, or that we need to get more words onto the stack)
# Check for left-arc, if yes carry out
# Check for right-arc, elif yes, carry out: your job is to fill in this portion of the code!
# else, perform a shift
# hints
# -checking against conditions for the right arc operation is fairly similar to the left arc check
#  with the order of stack1/stack2 swapped
# -to check if all the dependents of s1 have been assigned, you can
#  look for the presence of its children in dep_arcs
# -don't forget to update dep_arc after you called your perform_arc function

def tree_to_actions(wbuffer, stack, arcs, deps):

    # Initialize return lists
    
    # A list of 3-element tuples of lists
    # [(wbuffer1, stack1, arcs1), (wbuffer2, stack2, arcs2), ...]
    # Keeps track of the states at each step
    configurations=[]

    # gold_transitions:
    # A list of action strings, e.g ["SHIFT", "LEFTARC_nsubj"]
    # Keeps track of the actions at each step
    gold_transitions=[]

    # Keeps track of the dependents of each word in the tree
    # that have already been assigned
    dep_arcs = {}
    
    while len(wbuffer) >= 0:
        # Check for base-cases

        # firstly, check if we have translated all the tree to transition instructions
        if len(wbuffer) == 0 and len(stack) == 1 and stack[0] == 0:
            return configurations, gold_transitions

        # also, if there are fewer than 2 words on the stack
        # and more than 0 left on the buffer, 
        # we need to perform a shift operation
        if len(stack) < 2 and len(wbuffer) > 0:
            # shift operations
            perform_shift(wbuffer, stack, arcs, configurations, gold_transitions)
            continue

        # grab s1 and s2
        stack1 = stack[-1]
        stack2 = stack[-2]

        # check against conditions for left arc operation
        if stack1 in deps and (stack1, stack2) in deps[stack1]:
            # perform left arc
            perform_arc("LEFT", deps[stack1][(stack1, stack2)], wbuffer, stack, arcs, configurations, gold_transitions)
            # update dep_arcs
            dep_arcs[stack2] = 1

        # BEGIN SOLUTION  
        # check against conditions for right arc operation
        elif stack2 in deps and (stack2, stack1) in deps[stack2]:
            # check if all the dependents of s1 have been assigned
            if stack1 not in deps or all([dep[1] in dep_arcs for dep in deps[stack1].keys()]):
                # perform right arc
                perform_arc("RIGHT", deps[stack2][(stack2, stack1)], wbuffer, stack, arcs, configurations, gold_transitions)
                # update dep_arcs
                dep_arcs[stack1] = 1
            else:
                # perform shift
                perform_shift(wbuffer, stack, arcs, configurations, gold_transitions)
        # END SOLUTION

        # perform shift
        else:
            perform_shift(wbuffer, stack, arcs, configurations, gold_transitions)
    
    return configurations, gold_transitions

In [35]:

"""
objective: sanity check the tree_to_actions() function
"""
def sanity_check_tree_to_actions():

    # Setup for invoking tree_to_actions 
    wbuffer = [9, 8, 7, 6, 5, 4, 3, 2, 1] 
    stack = [0]
    arcs = []
    deps = {5: {(5, 9): 'punct', (5, 8): 'obl', (5, 4): 'advmod', (5, 3): 'aux:pass', (5, 2): 'nsubj:pass'},
            8: {(8, 7): 'det', (8, 6): 'case'}, 0: {(0, 5): 'root'}, 2: {(2, 1): 'nmod:poss'}}

    tree_to_actions(wbuffer, stack, arcs, deps)

    # After tree_to_actions
    assert wbuffer == [], "The result for wbuffer is not correct"
    assert stack == [0], "The result for stack is not correct"
    assert arcs == [('nmod:poss', 2, 1), ('advmod', 5, 4), ('aux:pass', 5, 3), ('nsubj:pass', 5, 2), 
                     ('det', 8, 7), ('case', 8, 6), ('obl', 5, 8), ('punct', 5, 9), ('root', 0, 5)], \
                        "The result for arcs is not correct"
    assert deps == {5: {(5, 9): 'punct', (5, 8): 'obl', (5, 4): 'advmod', (5, 3): 'aux:pass', (5, 2): 'nsubj:pass'}, 
                     8: {(8, 7): 'det', (8, 6): 'case'}, 0: {(0, 5): 'root'}, 2: {(2, 1): 'nmod:poss'}}, \
                    "The result for deps is not correct"
    print("You cleared the sanity check for tree_to_actions()!")   
     
sanity_check_tree_to_actions()    

shift attempted!
right arc attempted!
shift attempted!
right arc attempted!
right arc attempted!
You cleared the sanity check for tree_to_actions()!
